src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java

```diff
@@ -53,120 +53,141 @@ public ConfigBasedBeanDefinition(Object configInstance, Method beanMethod) {
+     * @return {@code true} if current {@link ConfigBasedBeanDefinition#instance} is not null, {@code false} otherwise.
      */
     @Override
-    public Object instance(BeanDefinition... dependencies) {
-        if (instance != null) {
-            log.debug("Getting existing '{}' bean instance", name);
-            return instance;
+    public boolean isInstantiated() {
+        return Objects.nonNull(instance);
+    }
+
+    /**
+     * See {@link BeanDefinition#instantiate(BeanDefinition...)}
+     *
+     * @param dependencies {@link BeanDefinition} objects that are treated as required dependencies for
+     *                     instantiating a bean from current {@link BeanDefinition}
+     */
+    @Override
+    public void instantiate(BeanDefinition... dependencies) {
+        if (!isInstantiated()) {
+            log.debug("Creating new instance of bean '{}'", name);
+            instance = createInstance(dependencies);
         }
-        return createInstance(dependencies);
+    }
+
+    /**
+     * See {@link BeanDefinition#getInstance()}.
+     *
+     * @return bean instance of current {@link BeanDefinition}.
+     * @throws NullPointerException when {@link ConfigBasedBeanDefinition#instance} is null.
+     */
+    @Override
+    public Object getInstance() {
+        return Objects.requireNonNull(instance, "Instance of %s has not been created yet".formatted(name));
+    }
+
+    private String resolveName(Method beanMethod) {
+        Bean annotation = beanMethod.getAnnotation(Bean.class);
+        String beanName = annotation.name();
+        if (beanName.isEmpty()) {
+            return beanMethod.getName();
+        }
+        return beanName;
+    }
+
+    private Class<?> getType(Method beanMethod) {
+        return beanMethod.getReturnType();
+    }
+
+    private Map<String, Class<?>> resolveDependencies(Method beanMethod) {
+        return Arrays.stream(beanMethod.getParameters())
+                .collect(toMap(this::resolveParameterName, Parameter::getType));
     }
 
     private Object createInstance(BeanDefinition... dependencies) {
         try {
-            log.debug("Initializing '{}' bean instance", name);
-            instance = getInstance(dependencies);
+            Object createdInstance = doCreateInstance(dependencies);
             log.debug("'{}' bean was instantiated", name);
-            return instance;
-
+            return createdInstance;
         } catch (Exception e) {
             throw new BeanInstanceCreationException("'%s' bean can't be instantiated".formatted(name), e);
         }
     }
 
-    private Object getInstance(BeanDefinition... dependencies) throws InvocationTargetException, IllegalAccessException {
-        Object[] args = Arrays.stream(dependencies)
-                              .map(BeanDefinition::instance)
-                              .toArray();
+    private Object doCreateInstance(BeanDefinition... dependencies) throws InvocationTargetException, IllegalAccessException {
```

Comment:
Could you please add a test that would make sure it sorts arguments in the expected order?

Reply:
will do